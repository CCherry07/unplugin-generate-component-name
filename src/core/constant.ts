export const EXPORT_HELPER_ID = '\0plugin-vue:export-helper'
