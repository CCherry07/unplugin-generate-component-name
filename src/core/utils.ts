import { basename, dirname, extname } from "path"
import type { VueQuery, FilteringRules } from "../types"

export function parseVueRequest(id: string): {
  filename: string
  query: VueQuery
} {
  const [filename, rawQuery] = id.split(`?`, 2)
  const query = Object.fromEntries(new URLSearchParams(rawQuery)) as VueQuery
  if (query.vue != null) {
    query.vue = true
  }
  if (query.index != null) {
    query.index = Number(query.index)
  }
  if (query.raw != null) {
    query.raw = true
  }
  if (query.url != null) {
    query.url = true
  }
  if (query.scoped != null) {
    query.scoped = true
  }
  return {
    filename,
    query,
  }
}

export const getComponentName = ({ filters, filename, attrs }: {
  filters?: FilteringRules
  filename: string
  attrs: Record<string, any>
}) => {
  const geComponentName = filters?.find(({ filter }) => filter(filename))?.geComponentName
  const originalName = basename(filename).replace(extname(filename), '')
  return geComponentName
    ? geComponentName({
      filePath: filename,
      originalName,
      attrName: attrs.name,
      dirname: basename(dirname(filename))
    })
    : attrs.name ?? originalName
}
