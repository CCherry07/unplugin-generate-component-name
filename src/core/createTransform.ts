import { parse as vueParse, compileScript } from '@vue/compiler-sfc'
import MagicString from 'magic-string'
import type { Filters } from '../types'
// @ts-ignore
import traverse from '@babel/traverse'
import { gte } from "semver"
import { getComponentName, parseVueRequest } from "./utils"

export const createTransform = (vueVersion?: string, filters?: Filters) => {
  return (code: string, id: string) => {
    const { filename } = parseVueRequest(id)
    const { descriptor } = vueParse(code, {
      filename: filename,
      ignoreEmpty: true
    })
    if (descriptor.script || descriptor.scriptSetup) {
      const { scriptAst, scriptSetupAst, loc, attrs } = compileScript(descriptor, { id });
      if (!scriptSetupAst && !scriptAst) return
      let hasNameProperty = false
      let isHandle = false
      const s = new MagicString(code);
      traverse({
        "type": "Program",
        "sourceType": "module",
        body: [...scriptAst ?? [], ...scriptSetupAst ?? []]
      }, {
        noScope: true,
        ExportDefaultDeclaration(path: { node: { declaration: { type: string; properties: any[] } } }) {
          if (hasNameProperty) { return }
          if (path.node.declaration.type === 'ObjectExpression') {
            hasNameProperty = path.node.declaration.properties.some(
              property => property.key.name === 'name'
            );
          }
        },
        CallExpression(path: any) {
          if (hasNameProperty) { return }
          const callExpr = path.node;
          if (callExpr.callee.type === "Identifier" && callExpr.callee.name === "defineOptions") {
            const arg = callExpr.arguments?.[0];
            const componentName = getComponentName({ filters, filename, attrs })
            if (arg && arg.type === "ObjectExpression") {
              for (const property of arg.properties) {
                if (property.key.name === "name" && property.value.type === "StringLiteral") {
                  hasNameProperty = true;
                }
              }
              if (!hasNameProperty) {
                const defineOptionsCode = s.slice(callExpr.start + loc.start.offset, callExpr.end + loc.start.offset)
                const startPos = defineOptionsCode.indexOf('{') + 1;
                s.appendLeft(callExpr.start + loc.start.offset + startPos, `name:'${componentName}',`);
                code = s.toString();
              }
            } else {
              const newCall = `defineOptions({name: "${componentName}"});\n`;
              s.overwrite(callExpr.start + loc.start.offset, callExpr.end + loc.start.offset, newCall);
              code = s.toString();
            }
            isHandle = true;
          }
        },
      })

      if (isHandle) {
        return {
          code,
          map: s.generateMap({
            hires: true
          })
        }
      }

      if (vueVersion && !hasNameProperty) {
        const componentName = getComponentName({ filters, filename, attrs })
        let lastImportEnd = 0;
        let defineOptionsExist = false
        traverse({
          "type": "Program",
          "sourceType": "module",
          body: [...scriptSetupAst ?? []],
        }, {
          noScope: true,
          ImportDeclaration(path: any) {
            lastImportEnd = path.node.end
            const specifiers = path.node.specifiers;
            for (const specifier of specifiers) {
              if (specifier.local.name === "defineOptions") {
                defineOptionsExist = true
              }
            }
          }
        })
        if (defineOptionsExist) {
          const newCall = `\ndefineOptions({ name: "${componentName}" }); \n`;
          s.appendLeft(loc.start.offset + lastImportEnd, newCall);
          code = s.toString();
        } else if (gte(vueVersion, '3.3.0')) {
          const newImport = `\nimport { defineOptions } from 'vue'; \n`;
          const newCall = `defineOptions({ name: "${componentName}" }); \n`;
          s.appendLeft(loc.start.offset + lastImportEnd, newImport + newCall);
          code = s.toString();
        } else if (scriptAst) {
          let ExportDefaultExist = false
          const regExp = /<\/script>/g;
          const matches = Array.from(code.matchAll(regExp));
          const [index] = matches.map(match => match.index).filter(idx => idx !== loc.end.offset);
          traverse({
            "type": "Program",
            "sourceType": "module",
            body: [...scriptAst ?? []],
          }, {
            noScope: true,
            ExportDefaultDeclaration(path: any) {
              ExportDefaultExist = true
              if (path.node.declaration.type === 'ObjectExpression') {
                const ExportDefaultIndex = code.indexOf('export default')
                s.appendLeft(code.slice(ExportDefaultIndex).indexOf('{') + ExportDefaultIndex + 1, `name:'${componentName}',`)
                code = s.toString()
              }
            }
          })
          if (!ExportDefaultExist && index) {
            const newExport = ` export default {
              name: "${componentName}",
            }; \n`
            s.appendLeft(index, newExport)
            code = s.toString()
          }
        } else {
          const newExport = `
            <script lang='${attrs.lang}'>
              export default {
                name: "${componentName}",
              }; \n
              </script> \n`;
          s.appendLeft(0, newExport);
          code = s.toString();
        }
        return {
          code,
          map: s.generateMap({
            hires: true
          })
        }
      }
    }
  }
}
