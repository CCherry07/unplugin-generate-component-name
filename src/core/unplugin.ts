import { resolve } from "node:path"
import fs from "fs/promises"
import { parse as vueParse, compileScript } from '@vue/compiler-sfc'
import { createUnplugin } from "unplugin"
import MagicString from 'magic-string'
import { createFilter } from "@rollup/pluginutils"
// @ts-ignore
import traverse from '@babel/traverse'
import { gte, minVersion } from "semver"
import { Options } from "../types"
import { getComponentName, parseVueRequest } from "./utils"

export default createUnplugin((options: Options = {}) => {
  const filter = createFilter(
    options.include || ["**/index.vue"],
    options.exclude || [/[\\/]node_modules[\\/]/, /[\\/]\.git[\\/]/, /[\\/]\.nuxt[\\/]/],
  )
  let vueVersion: string | undefined
  return {
    name: "GeComponentName",
    enforce: 'pre',
    async buildStart() {
      const packageJsonPath = resolve(process.cwd(), 'package.json');
      await fs.readFile(packageJsonPath).then(fileData => {
        const dependencies = JSON.parse(fileData.toString()).dependencies;
        vueVersion = minVersion(dependencies['vue'])?.version
      });
    },
    transformInclude(id) {
      const { filename, query } = parseVueRequest(id)
      if (query.raw || query.url) return
      if (!filter(filename) && !query.vue) return
      return true
    },
    async transform(code, id) {
      const { filename } = parseVueRequest(id)
      const { descriptor } = vueParse(code, {
        filename: filename,
        ignoreEmpty: true
      })
      if (descriptor.script || descriptor.scriptSetup) {
        const { scriptAst, scriptSetupAst, loc, attrs } = compileScript(descriptor, { id });
        if (!scriptSetupAst && !scriptAst) return
        let hasNameProperty = false
        let isHandle = false
        const s = new MagicString(code);
        traverse({
          "type": "Program",
          "sourceType": "module",
          body: [...scriptAst ?? [], ...scriptSetupAst ?? []]
        }, {
          noScope: true,
          ExportDefaultDeclaration(path: { node: { declaration: { type: string; properties: any[] } } }) {
            if (hasNameProperty) { return }
            if (path.node.declaration.type === 'ObjectExpression') {
              hasNameProperty = path.node.declaration.properties.some(
                property => property.key.name === 'name'
              );
            }
          },
          CallExpression(path: any) {
            if (hasNameProperty) { return }
            const callExpr = path.node;
            if (callExpr.callee.type === "Identifier" && callExpr.callee.name === "defineOptions") {
              const arg = callExpr.arguments?.[0];
              if (arg && arg.type === "ObjectExpression") {
                for (const property of arg.properties) {
                  if (property.key.name === "name" && property.value.type === "StringLiteral") {
                    hasNameProperty = true;
                  }
                }
                if (!hasNameProperty) {
                  const defineOptionsCode = s.slice(callExpr.start + loc.start.offset, callExpr.end + loc.start.offset)
                  const startPos = defineOptionsCode.indexOf('{') + 1;
                  s.appendLeft(callExpr.start + loc.start.offset + startPos, `name:'${getComponentName({
                    geComponentName: options.geComponentName,
                    filename,
                    attrs
                  })}',`);
                  code = s.toString();
                }
              } else {
                const newCall = `defineOptions({name: "${getComponentName({
                  geComponentName: options.geComponentName,
                  filename,
                  attrs
                })}"});\n`;
                s.overwrite(callExpr.start + loc.start.offset, callExpr.end + loc.start.offset, newCall);
                code = s.toString();
              }
              isHandle = true;
            }
          },
        })

        if (isHandle) {
          return {
            code,
            map: s.generateMap({
              hires: true
            })
          }
        }

        if (vueVersion && !hasNameProperty) {
          const componentName = getComponentName({ geComponentName: options.geComponentName, filename, attrs })
          let lastImportEnd = 0;
          let defineOptionsExist = false
          traverse({
            "type": "Program",
            "sourceType": "module",
            body: [...scriptSetupAst ?? []],
          }, {
            noScope: true,
            ImportDeclaration(path: any) {
              lastImportEnd = path.node.end
              const specifiers = path.node.specifiers;
              for (const specifier of specifiers) {
                if (specifier.local.name === "defineOptions") {
                  defineOptionsExist = true
                }
              }
            }
          })
          if (defineOptionsExist) {
            const newCall = `\ndefineOptions({ name: "${componentName}" }); \n`;
            s.appendLeft(loc.start.offset + lastImportEnd, newCall);
            code = s.toString();
          } else if (gte(vueVersion, '3.3.0')) {
            const newImport = `\nimport { defineOptions } from 'vue'; \n`;
            const newCall = `defineOptions({ name: "${componentName}" }); \n`;
            s.appendLeft(loc.start.offset + lastImportEnd, newImport + newCall);
            code = s.toString();
          } else if (scriptAst) {
            let ExportDefaultExist = false
            const regExp = /<\/script>/g;
            const matches = Array.from(code.matchAll(regExp));
            const [index] = matches.map(match => match.index).filter(idx => idx !== loc.end.offset);
            traverse({
              "type": "Program",
              "sourceType": "module",
              body: [...scriptAst ?? []],
            }, {
              noScope: true,
              ExportDefaultDeclaration(path: any) {
                ExportDefaultExist = true
                if (path.node.declaration.type === 'ObjectExpression') {
                  const ExportDefaultIndex = code.indexOf('export default')
                  s.appendLeft(code.slice(ExportDefaultIndex).indexOf('{') + ExportDefaultIndex + 1, `name:'${getComponentName({
                    geComponentName: options.geComponentName,
                    filename,
                    attrs
                  })}',`)
                  code = s.toString()
                }
              }
            })
            if (!ExportDefaultExist && index) {
              const newExport = ` export default {
                name: "${componentName}",
              }; \n`
              s.appendLeft(index, newExport)
              code = s.toString()
            }
          } else {
            const newExport = `
              <script lang='${attrs.lang}'>
                export default {
                  name: "${componentName}",
                }; \n
                </script> \n`;
            s.appendLeft(0, newExport);
            code = s.toString();
          }
          return {
            code,
            map: s.generateMap({
              hires: true
            })
          }
        }
      }
    },
  }
})
