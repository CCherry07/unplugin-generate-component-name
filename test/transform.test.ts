import { describe, expect, it } from 'vitest'
import ComponentA from "./components/ComponentA/index.vue?raw"
import ComponentB from "./components/ComponentB/index.vue?raw"
import TestSetupName from "./components/test-setup-name.vue?raw"
import { createTransform } from "../src/core/createTransform"

describe('The behavior of transform in Vue 3.3.0 and above.', () => {
  const transform = createTransform("3.3.0")
  it('Component name is dirname', () => {
    const code = transform(ComponentA, 'components/ComponentA/index.vue')?.code
    expect(code).toContain('defineOptions({ name: "ComponentA" });')
    expect(code).toMatchSnapshot()
  })

  it('setup extend name', () => {
    const code = transform(TestSetupName, 'components/test-setup-name.vue')?.code
    expect(code).toContain('defineOptions({ name: "NameForeSetup" });')
  })

  it('when define options exists', () => {
    const code = transform(ComponentB, 'components/ComponentB/index.vue')?.code
    expect(code).toContain('inheritAttrs: false')
    expect(code).toContain("name:'ComponentB'")
    expect(code).toMatchInlineSnapshot(`
      "<template>
        <div>
          {{ b }}
        </div>
      </template>

      <script setup lang="ts">
      import { defineOptions } from "vue"
      defineOptions({name:'ComponentB',
        inheritAttrs: false
      })
      const b = ('component B')
      </script>
      "
    `)
  })
})
