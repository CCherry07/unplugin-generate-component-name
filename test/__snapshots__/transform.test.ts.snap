// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`The behavior of transform in Vue 3.3.0 and above. > Component name is dirname 1`] = `
"<template>
  <div>
    {{ a }}
    <ComponentB/>
  </div>
</template>
<script setup lang="ts">
import ComponentB from "../ComponentB/index.vue";
import { defineOptions } from 'vue'; 
defineOptions({ name: "ComponentA" }); 

const a = ('component A')
</script>
"
`;
